Spigot-Server-Patches/0613-Fix-item-position-desync.patch

```diff
@@ -0,0 +1,58 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: BillyGalbreath <blake.galbreath@gmail.com>
+Date: Tue, 8 Dec 2020 20:24:52 -0600
+Subject: [PATCH] Fix item position desync
+
+This fixes item position desync (MC-4) by running the item coordinates
+through the encryption/decryption methods of the packet that causes the
+precision loss, which forces the server to lose the same precision as
+the client keeping them in sync.
+
+diff --git a/src/main/java/com/destroystokyo/paper/PaperConfig.java b/src/main/java/com/destroystokyo/paper/PaperConfig.java
+index 6ab0a51b421dc696aec292f2db955311c9cc31e3..c70b80ab32d7bac66f69a08f25726e73114bb7fe 100644
+--- a/src/main/java/com/destroystokyo/paper/PaperConfig.java
++++ b/src/main/java/com/destroystokyo/paper/PaperConfig.java
+@@ -462,4 +462,9 @@ public class PaperConfig {
+     private static void trackPluginScoreboards() {
+         trackPluginScoreboards = getBoolean("settings.track-plugin-scoreboards", false);
+     }
++
++    public static boolean fixItemPositionDesync = true;
++    private static void fixItemPositionDesync() {
++        fixItemPositionDesync = getBoolean("settings.fix-item-position-desync", fixItemPositionDesync);
++    }
+ }
+diff --git a/src/main/java/net/minecraft/server/EntityTrackerEntry.java b/src/main/java/net/minecraft/server/EntityTrackerEntry.java
+index 4efc40c01ec12b80bd7cf9d35cf0ea0df973baf7..a0bd864290f11abf2a7be9c70f21406529f21eeb 100644
+--- a/src/main/java/net/minecraft/server/EntityTrackerEntry.java
++++ b/src/main/java/net/minecraft/server/EntityTrackerEntry.java
+@@ -133,6 +133,12 @@ public class EntityTrackerEntry {
+                 double vec3d_dz = this.tracker.locZ() - 2.44140625E-4D*(this.zLoc);
+                 boolean flag1 = (vec3d_dx * vec3d_dx + vec3d_dy * vec3d_dy + vec3d_dz * vec3d_dz) >= 7.62939453125E-6D;
+                 // Paper end - reduce allocation of Vec3D here
++                // Paper start - fix MC-4
++                if (com.destroystokyo.paper.PaperConfig.fixItemPositionDesync && this.tracker instanceof EntityItem) {
```

Comment:
doing this in tracker feels odd, itll be done multiple times for players.

Shouldnt this be done in the setPosition methods that update the  x/y/z directly?

Reply:
also, im thinking we should maybe just inline the logic to avoid creating a vec3d just to do this.